#2: added initial project script, .gitignore and README.md files (#3)
exasol/row-level-security: doc/user_guide/user_guide.md

# Row Level Security

> Row Level Security is an extension of Virtual Schemas Exasol dialect. It allows database administrators to set an access to a table's row basing on user's roles and user's names.

Reviewer: Implementation detail.
Better:
```suggestion
Row-Level Security (short "RLS") is a security mechanism based on Exasol's Virtual Schemas. It allows database administrators to set an access to a table's row basing on user's roles and user's names.

RLS only supports Exasol databases. That means you cannot use RLS between Exasol and a 3rd-party data source.
```

Author: done

> You don't have to install any JDBC driver, because it is already installed in the Exasol database and also included in the JAR of the JDBC adapter.

Reviewer: ```suggestion
The RLS installation package contains everything you need to extend an existing Exasol installation with row-level security.
```

The part about the driver was an implementation detail. Also at this point in the guide users don't have the base information to understand, why RLS requires a database driver at all.

Author: done

## Row Level Security Tables Preparations

> ### Roles security

Reviewer: ```suggestion
### Role-based security
```

Also fix other occurrences.

Author: done

> 

Reviewer: Explain first what "role-based security" means.

Author: done

> 1. In case you want to use roles security, you must add an additional column EXA_ROW_ROLES DECIMAL(20,0) to the tables you want to secure. 

Reviewer: Remove the numbers please.

Reviewer: Also wrap identifiers in backticks.

Author: done

> Example:

Reviewer: Add empty lines on top and below.

Author: done

```sql
CREATE OR REPLACE TABLE row_level_security_test_schema.rls_sales 
    (  
    order_id DECIMAL(18,0),  
    customer VARCHAR(50),  
    product VARCHAR(100),  
    quantity DECIMAL(18,0),
    exa_row_roles DECIMAL(20,0)  
    ); 
```

> Then fill the EXA_ROW_ROLES column with the roles.

Reviewer: Use case consistently. I suggests upper case everywhere since that is what the database automatically makes out of unquoted identifiers.

Author: done

> 

Reviewer: Add a note here that later versions of RLS will provide UDF functions that make administration of RLS more user-friendly.

Author: done

2. Create a configure table with user roles and fill it with the roles:
```sql
CREATE OR REPLACE TABLE row_level_security_test_schema.exa_rls_users  
    (  
    exa_user_name VARCHAR(200),
    exa_role_mask DECIMAL(20,0)  
    );  

INSERT INTO row_level_security_test_schema.exa_rls_users VALUES 
('RLS_USR_1', NULL),
('RLS_USR_2', 1),
('RLS_USR_3', 3),
('RLS_USR_4', 15),
('SYS', 15);
```

> ### Tenants security

Reviewer: ```suggestion
### Tenant-based security
```

Reviewer: Add a general introduction: What is a tenant? Data can only belong to exactly one tenant or be public.

Author: done

If you want to use tenant security, you must add an additional column EXA_ROW_TENANTS VARCHAR(128) to the tables you want to secure. 
Example:
```sql
CREATE OR REPLACE TABLE row_level_security_test_schema.rls_sales 
    (  
    order_id DECIMAL(18,0),  
    customer VARCHAR(50),  
    product VARCHAR(100),  
    quantity DECIMAL(18,0),

>     exa_row_tenants VARCHAR(128)  

Reviewer: Please rename to `exa_row_tenant` to emphasize the fact that there can only be one. Also in the code of course.

Author: After automatic renaming some logic is broken. I created a ticket for this refactoring.

    );     
```

> Then fill the EXA_ROW_TENANTS column with the user's names.

Reviewer: ```suggestion
For each row define which tenant it belongs to.
```

Author: done

Upload the latest available release of [Row Level Security](https://github.com/exasol/row-level-security/releases) to Bucket FS.

Then create a schema or use an existing one to hold the adapter script.

```sql
CREATE SCHEMA TEST_SCHEMA;
```

The SQL statement below creates the adapter script, defines the Java class that serves as entry point and tells the UDF framework where to find the libraries (JAR files) for Virtual Schema and database driver.

```sql
CREATE JAVA ADAPTER SCRIPT TEST_SCHEMA.TEST_ADAPTER_SCRIPT AS
    %scriptclass com.exasol.adapter.RequestDispatcher;
    %jar /buckets/<BFS service>/<bucket>/row-level-security-<version>-all-dependencies.jar;
/
```

### Defining a Named Connection

```sql
CREATE CONNECTION EXASOL_JDBC_CONNECTION 
TO 'jdbc:exa:<host>:<port>' 
USER '<user>' 
IDENTIFIED BY '<password>';
```

### Creating a Virtual Schema

```sql
CREATE VIRTUAL SCHEMA VIRTUAL_SCHEMA_RLS 
    USING TEST_SCHEMA.TEST_ADAPTER_SCRIPT WITH
    SQL_DIALECT     = 'EXASOL_RLS'
    CONNECTION_NAME = 'EXASOL_JDBC_CONNECTION'
    SCHEMA_NAME     = '<schema name>'
    IMPORT_FROM_EXA = 'true'
    EXA_CONNECTION_STRING = '<host>:<port>';
```

### Defining a Named Connection

> Define the connection to the other Exasol cluster as shown below.

Reviewer: ```suggestion
Define the connection to the other Exasol instance as shown below.
```

Author: done

```sql
CREATE CONNECTION EXASOL_CONNECTION 
TO 'jdbc:exa:<host>:<port>' 
USER '<user>' 
IDENTIFIED BY '<password>';
```

### Creating a Virtual Schema

```sql
CREATE VIRTUAL SCHEMA <virtual schema name> 
    USING ADAPTER.JDBC_ADAPTER 
    WITH
    SQL_DIALECT     = 'EXASOL_RLS'
    CONNECTION_NAME = 'EXASOL_CONNECTION'
    SCHEMA_NAME     = '<schema name>';
```

### Limitations

> - RLS Virtual Schema do not support JOIN capabilities.

Reviewer: - RLS only works with Exasol as source and destination of the Virtual Schema.

Author: done
